File: 380-insert-delete-getrandom-o1/insert-delete-getrandom-o1.py
class RandomizedSet:
    def __init__(self):
        self.my_dict= {}
        self.keys = []
        

    def insert(self, val: int) -> bool:
        if val not in self.my_dict:
            self.my_dict[val] = len(self.keys) 
            self.keys.append(val)
            return True
        return False

    def remove(self, val: int) -> bool:
        if val in self.my_dict:
            # find index of val from dict 
            idx = self.my_dict[val]
            # find last element from the list
            last_val = self.keys[-1]

            #swap val with last element
            #-so the last element is in the place of val and remove the last element 
            self.keys[idx] = last_val
            #update the dict so that the last val has the index of val 
            self.my_dict[last_val] = idx

            # remove last element 
            self.keys.pop()
            del self.my_dict[val]
            return True
        return False
        

    def getRandom(self) -> int:
        if self.keys :
            return random.choice(self.keys)
        return []
        


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()

File: 380-insert-delete-getrandom-o1/insert-delete-getrandom-o1.js

var RandomizedSet = function() {
    this.nums = []
    this.my_dict = new Map()
    
};

RandomizedSet.prototype.insert = function(val) {
    if (this.my_dict.has(val)) return false

    
    // insert to a map => map.set(key, value)
    this.my_dict.set(val,this.nums.length)
    this.nums.push(val)

    return true
};

RandomizedSet.prototype.remove = function(val) {
    if(!this.my_dict.has(val)) return false
    
    // get the last element in the list and the index of val 
    const index_of_val = this.my_dict.get(val)
    const last_element = this.nums[this.nums.length-1]

    //swap - put last element inplace of val
    this.nums[index_of_val] = last_element
    this.my_dict.set(last_element,index_of_val)
    
    // remove the last element from nums and remove val from the map
    this.nums.pop()
    this.my_dict.delete(val)
    return true
};

RandomizedSet.prototype.getRandom = function() {
    const rand_index = Math.floor(Math.random()*this.nums.length)
    return this.nums[rand_index]
};
